### PR TITLE
Fix tied oracle embedding export

### LINKED ISSUE #1: [BUG] Materialize tied oracle embeddings during checkpoint export
## Bug Description

When a model is trained with oracle_embedding_mode=tie under DeepSpeed ZeRO, text_emb is used and updated for oracle input while oracle_emb can remain at its initial value in the consolidated checkpoint. tools.clean_moshi currently converts and saves that state without applying the training-time tie contract. The resulting inference checkpoint can therefore contain a stale oracle embedding.

## Steps to Reproduce

1. Fine-tune with oracle_embedding_mode=tie using the DeepSpeed ZeRO checkpoint path.
2. Consolidate the checkpoint with tools.zero_to_fp32.
3. Convert it with tools.clean_moshi.
4. Compare the text_emb and oracle_emb state in the cleaned checkpoint.

## Expected Behavior

For a model trained in tie mode, the exporter should copy the complete learned text_emb state into oracle_emb after ZeRO consolidation and validate exact equality before saving the inference checkpoint.

For a model trained in separate mode, the separately trained oracle_emb state should remain unchanged.

## Actual Behavior

tools.clean_moshi does not currently accept or apply the training-time oracle embedding mode, so a stale oracle_emb can be serialized for tie-mode training.

## Environment

- Repository revision: current main
- Training path: DeepSpeed ZeRO checkpoint followed by zero_to_fp32 and clean_moshi

## Additional Information

The default and documented training mode is separate, so the default public workflow and the currently published KAME checkpoint are not affected by this tie-mode export issue. The inference runtime is also not the source of the problem.

## Possible Solution

Require the training-time oracle embedding mode during clean export. In tie mode, materialize oracle_emb from text_emb after loading the consolidated fp32 model, validate exact key, shape, dtype, finite-value, and tensor equality, then validate again before serialization. In separate mode, preserve oracle_emb unchanged.

### PR DESCRIPTION
**Pull request by yagumana**
## Summary

- materialize `text_emb` into `oracle_emb` for tie-mode inference export after ZeRO consolidation, with strict key, shape, dtype, finiteness, and equality validation before serialization
- read `oracle_embedding_mode` from the training run's `config.json` by default and treat that metadata as the source of truth
- use `--oracle_embedding_mode` as a metadata cross-check or legacy fallback only, rejecting mismatches and invalid metadata before model loading
- preserve separately trained oracle embeddings unchanged in `separate` mode
- make `tools` an explicit package so Mypy resolves `tools.*` modules consistently
- document the export contract and add focused regression coverage

## Context

In `tie` mode, training uses the learned `text_emb` for oracle input, but ZeRO consolidation reconstructs trainable and frozen parameters through different checkpoint paths. A model-side copy before checkpointing is therefore not sufficient to guarantee that the consolidated fp32 checkpoint contains the learned oracle table. Without an export-side materialization step, `clean_moshi` could serialize a stale `oracle_emb`.

The training run already records `oracle_embedding_mode` in its root `config.json`. This PR now uses that metadata as the authoritative mode instead of relying solely on a manually supplied export flag. `clean_moshi` discovers the standard parent config automatically, accepts an explicit config path for nonstandard layouts, cross-checks an optional CLI mode, and permits CLI-only fallback only for legacy checkpoints without metadata. Invalid metadata, malformed JSON, explicit missing paths, and metadata/CLI mismatches fail before model loading.

Fixes #1

## Validation

- `uv run pytest -q` — 65 passed
- `uv run ruff check .`
- `uv run ruff format --check .`
- `uv run python -m tools.clean_moshi --help`
- `uv run python -m tools.generate_oracle_from_text --help`
- `git diff --check`
- [GitHub Actions Code Quality run 31079767473](https://github.com/SakanaAI/kame_finetune/actions/runs/31079767473) — passed

Mypy now resolves all `tools.*` modules consistently and checks 31 source files instead of stopping at a duplicate-module error. It still reports the repository's 38 pre-existing, non-blocking errors in seven unchanged files; none are in this PR's four changed files.

### CODE REVIEW

File: tools/clean_moshi.py [hunk 1]
```diff
@@ -12,13 +12,74 @@
...
 
 
+def _validate_tied_oracle_embedding(model, *, context: str) -> None:
+    text_state = model.text_emb.state_dict()
+    oracle_state = model.oracle_emb.state_dict()
+    if text_state.keys() != oracle_state.keys():
+        raise RuntimeError(
+            f"Tied oracle embedding state mismatch {context}: "
+            f"{sorted(text_state)} != {sorted(oracle_state)}"
+        )
```
[the remark this answers is not in the document]
Comment (yagumana) [reply tools/clean_moshi.py:110]: Updated the error formatting to sort text_state.keys() and oracle_state.keys() explicitly. This is a readability-only change; focused clean_moshi tests and Ruff checks pass. Addressed in 147fe45.

**Comment by Kuroki1931** [comment]
@yagumana 
The fix itself LGTM. One concern is that this design assumes whoever runs the export knows which mode (`tie` / `separate`) was used during training. If they don't know, or pass the wrong flag, no validation runs and a silently broken model gets exported. That is the same class of bug this PR fixes.

The obvious alternative would be to do the copy at save time on the training side, but I suspect that might be harder than it looks. If I understand correctly, with mixed-precision training the `text_emb` accessible at save time is the half-precision model copy, while the fp32 master weights are sharded inside the optimizer. So a save-time copy could leave `text_emb` and `oracle_emb` slightly inconsistent after `zero_to_fp32`, whereas the export side copies the fp32-restored values and avoids this. Gathering parameters under ZeRO-3 might also add some complexity. I have not verified this in detail though, so please correct me if I am wrong.

So here is a suggestion. Record `oracle_embedding_mode` in the existing metadata at training time (e.g. the config JSON), and have `clean_moshi` read it automatically by default. The `--oracle_embedding_mode` flag could stay as an explicit override for existing checkpoints that lack the metadata, and serve as a cross-check when the metadata is present, keeping backward compatibility.

**Comment by yagumana** [comment]
@Kuroki1931 
Thanks — I agree that the export should not rely solely on a manually supplied mode.

I updated the PR so that `clean_moshi` now reads `oracle_embedding_mode` from the training run’s root `config.json` by default, or from an explicitly provided `--training_config_path`. Valid metadata is treated as the source of truth. When `--oracle_embedding_mode` is also supplied, it acts as a cross-check, and any mismatch fails before the model is loaded. CLI-only mode is retained as a warned fallback for legacy checkpoints without the metadata, while malformed or invalid metadata fails closed.

I also kept the materialization step after ZeRO consolidation, as you suggested. Since the final FP32 state is reconstructed from the ZeRO checkpoint, a model-side copy before checkpointing is not a sufficiently reliable source of truth for the exported weights. In `tie` mode, the exporter therefore copies the consolidated `text_emb` state into `oracle_emb` and validates exact equality before serialization.

Regression coverage now includes metadata discovery, CLI/metadata mismatch handling, invalid-metadata failures, the legacy fallback path, tie-mode copy direction, and separate-mode preservation. All 65 tests and Ruff checks pass.

Thanks for catching this.

**Review by Kuroki1931 — APPROVED** [review]
